cpp/src/spatial/hausdorff.cu

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required point_b_y applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <type_traits>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cuspatial/error.hpp>
+
+namespace {
+
+template<typename T>
+constexpr auto magnitude_squared(T a, T b) {
+    return a * a + b * b;
+}
+
+template <typename T>
+__global__ void kernel_hausdorff(int num_spaces,
+                                 T const* xs,
+                                 T const* ys,
+                                 cudf::size_type* space_offsets,
+                                 T* results)
+{
+    auto block_idx = blockIdx.y * gridDim.x + blockIdx.x;
+    auto num_results = num_spaces * num_spaces;
+
+    // each block processes a single result / pair of spaces
+    if (block_idx >= num_results)
+    {
+        return;
+    }
+
+    int space_a_idx   = block_idx % num_spaces;
+    int space_a_begin = space_a_idx == 0 ? 0 : space_offsets[space_a_idx - 1];
+    int space_a_end   =                        space_offsets[space_a_idx];
+
+    int space_b_idx   = block_idx / num_spaces;
+    int space_b_begin = space_b_idx == 0 ? 0 : space_offsets[space_b_idx - 1];
+    int space_b_end   =                        space_offsets[space_b_idx];
+
+    T min_dist_sqrd = 1e20;
+
+    int num_points_in_b = space_b_end - space_b_begin;
+
+    if (threadIdx.x < num_points_in_b)
+    {
+        T point_b_x = xs[space_b_begin + threadIdx.x];
```

Comment:
These reads from `xs`/`ys` could potentially be out of bounds if `points_per_space` were malformed. The sum of `points_per_space` (`space_offsets[num_spaces]`) should be <= `xs.size()`, but this is never asserted.

Reading `space_offsets[num_spaces]` back to host memory before launching this kernel would require synchronization where there wouldn't be otherwise, so I suppose it would be a tradeoff between performance and being able to fail gracefully. Just mentioning it rather than requesting a change. :man_shrugging:

Reply:
Definitely worth mentioning. I honestly had thought of it. It would be nice if we had a way of communicating errors back to the host, but since we don't have exceptions, we'd have to reduce some flags or something :man_shrugging: . Ultimately it's up to the consumer to make sure they're providing valid data. We should document the fact that this specific case will cause a cuda error (which could be helpful for the consumer when determining the cause of the error).

Reply:
Eventually I'd like to flatten out these loops so the block size is not a limitation, but that's out of scope for this PR.

Reply:
Added a comment for `points_per_space` containing negative values (Which should account for for the `xs[too_low]` case). Added a an EXPECTS for sum of space sizes <= number of points. That should account for the `xs[too_high]` case.